### PR TITLE
Have matrix square root account for Rii=Rjj=0 to prevent NaN values

### LINKED ISSUE #700: spurious NaN values from matrix square root
As reported [on discourse](https://discourse.julialang.org/t/matrix-square-root-of-known-positive-semi-definite-matrix-does-not-exist):
```jl
A = [
0.9999999999999998	4.649058915617843e-16	-1.3149405273715513e-16	9.9959579317056e-17
-8.326672684688674e-16	1.0000000000000004	2.9280733590254494e-16	-2.9993900031619594e-16
9.43689570931383e-16	-1.339206523454095e-15	1.0000000000000007	-8.550505126287743e-16
-6.245004513516506e-16	-2.0122792321330962e-16	1.183061278035052e-16	1.0000000000000002
]

B = [
0.09648289218436859	0.023497875751503007	0.0	0.0
0.023497875751503007	0.045787575150300804	0.0	0.0
0.0	0.0	0.0	0.0
0.0	0.0	0.0	0.0
]
sqrt(A*B*A')
```
gives
```
4×4 Array{Complex{Float64},2}:
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
```
This is incorrect.  In fact, up to roundoff error, the correct answer is:
```
4×4 Array{Float64,2}:
 0.307265   0.0455076  0.0  0.0
 0.0455076  0.209085   0.0  0.0
 0.0        0.0        0.0  0.0
 0.0        0.0        0.0  0.0
```
Roundoff errors may give a complex result here (see also JuliaLang/julia#35057), but should not lead to `NaN` values.

### LINKED ISSUE #35058: spurious NaN values from matrix square root
As reported [on discourse](https://discourse.julialang.org/t/matrix-square-root-of-known-positive-semi-definite-matrix-does-not-exist):
```jl
A = [
0.9999999999999998	4.649058915617843e-16	-1.3149405273715513e-16	9.9959579317056e-17
-8.326672684688674e-16	1.0000000000000004	2.9280733590254494e-16	-2.9993900031619594e-16
9.43689570931383e-16	-1.339206523454095e-15	1.0000000000000007	-8.550505126287743e-16
-6.245004513516506e-16	-2.0122792321330962e-16	1.183061278035052e-16	1.0000000000000002
]

B = [
0.09648289218436859	0.023497875751503007	0.0	0.0
0.023497875751503007	0.045787575150300804	0.0	0.0
0.0	0.0	0.0	0.0
0.0	0.0	0.0	0.0
]
sqrt(A*B*A')
```
gives
```
4×4 Array{Complex{Float64},2}:
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
```
This is incorrect.  In fact, up to roundoff error, the correct answer is:
```
4×4 Array{Float64,2}:
 0.307265   0.0455076  0.0  0.0
 0.0455076  0.209085   0.0  0.0
 0.0        0.0        0.0  0.0
 0.0        0.0        0.0  0.0
```
Roundoff errors may give a complex result here (see also #35057), but should not lead to `NaN` values.

### PR DESCRIPTION
**Pull request by ssikdar1**
# Description

Tries to Fix JuliaLang/LinearAlgebra.jl#700

Add a check when `R[i,i]` and `R[j,j]` is 0 to not call `sylvester(0,0,-r)`

# Testing

```
julia> sqrt(A*B*A')
4×4 Array{Complex{Float64},2}:
     0.307265+0.0im     0.0455076+0.0im   2.29019e-16+0.0im  -2.01045e-16+0.0im
    0.0455076+0.0im      0.209085+0.0im  -2.37063e-16+0.0im  -7.04933e-17+0.0im
  2.29019e-16+0.0im  -2.37063e-16+0.0im   5.33599e-31+0.0im  -9.53187e-32+0.0im
 -2.01045e-16+0.0im  -7.04933e-17+0.0im  -9.53187e-32+0.0im   1.39738e-31+0.0im
```

Where previously it returned 

```
julia> sqrt(A*B*A')
4×4 Array{Complex{Float64},2}:
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
 NaN+NaN*im  NaN+NaN*im  NaN+NaN*im  NaN+NaN*im
```

### CODE REVIEW

File: stdlib/LinearAlgebra/src/triangular.jl [hunk 1]
```diff
@@ -2607,7 +2607,11 @@ function sqrt(A::UpperTriangular{T},::Val{realmatrix}) where {T,realmatrix}
             @simd for k = i+1:j-1
                 r -= R[i,k]*R[k,j]
             end
-            iszero(r) || (R[i,j] = sylvester(R[i,i],R[j,j],-r))
+            if iszero(R[i,i])  && iszero(R[j,j])
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/triangular.jl:2610]: extra space before `&&`

File: stdlib/LinearAlgebra/src/triangular.jl [hunk 2]
```diff
@@ -2607,7 +2607,11 @@ function sqrt(A::UpperTriangular{T},::Val{realmatrix}) where {T,realmatrix}
...
+                R[i,j] = 0
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/triangular.jl:2611]: Isn't `R[i,j]` already zero?

File: stdlib/LinearAlgebra/test/triangular.jl [hunk 3]
```diff
@@ -651,4 +651,20 @@ end
...
+       -6.245004513516506e-16        -2.0122792321330962e-16        1.183061278035052e-16        1.0000000000000002
+    ]
+B = [
+       0.09648289218436859        0.023497875751503007        0.0        0.0
+       0.023497875751503007        0.045787575150300804        0.0        0.0
+       0.0        0.0        0.0        0.0
+       0.0        0.0        0.0        0.0
+    ]
+C = sqrt(A*B*A')
+@test !any(isnan,C)
```
Comment (stevengj) [inline stdlib/LinearAlgebra/test/triangular.jl:668]: Better
```jl
@test C^2 ≈ A*B*A'
```
which will also fail if there are NaN values in `C`.

Comment (ssikdar1) [reply stdlib/LinearAlgebra/test/triangular.jl]: updated in 573d006

File: stdlib/LinearAlgebra/src/triangular.jl (hunk 1 above)
Comment (ssikdar1) [reply stdlib/LinearAlgebra/src/triangular.jl]: removed in 573d006

File: stdlib/LinearAlgebra/src/triangular.jl (hunk 2 above)
Comment (ssikdar1) [reply stdlib/LinearAlgebra/src/triangular.jl]: yes you're right, fixed the ifelse in 573d006

File: stdlib/LinearAlgebra/src/triangular.jl [hunk 4]
```diff
@@ -2607,7 +2607,9 @@ function sqrt(A::UpperTriangular{T},::Val{realmatrix}) where {T,realmatrix}
             @simd for k = i+1:j-1
                 r -= R[i,k]*R[k,j]
             end
-            iszero(r) || (R[i,j] = sylvester(R[i,i],R[j,j],-r))
+            if !(iszero(R[i,i]) && iszero(R[j,j]))
+                iszero(r) || (R[i,j] = sylvester(R[i,i],R[j,j],-r))
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/triangular.jl:2611]: Would be clearer to combine all of the `if` statements:
```jl
if !(iszero(r) || (iszero(R[i,i]) && iszero(R[j,j])))
    R[i,j] = sylvester(R[i,i],R[j,j],-r)
end
```

File: stdlib/LinearAlgebra/test/triangular.jl [hunk 5]
```diff
@@ -651,4 +651,20 @@ end
...
+       -6.245004513516506e-16        -2.0122792321330962e-16        1.183061278035052e-16        1.0000000000000002
+    ]
+B = [
+       0.09648289218436859        0.023497875751503007        0.0        0.0
+       0.023497875751503007        0.045787575150300804        0.0        0.0
+       0.0        0.0        0.0        0.0
+       0.0        0.0        0.0        0.0
+    ]
+C = sqrt(A*B*A')
+@test C^2 ≈ A*B*A'
```
Comment (stevengj) [inline stdlib/LinearAlgebra/test/triangular.jl:668]: I would declare these as locals with a `let` statement:
```jl
let A = [0.9999999999999998 4.649058915617843e-16 -1.3149405273715513e-16 9.9959579317056e-17; -8.326672684688674e-16 1.0000000000000004 2.9280733590254494e-16 -2.9993900031619594e-16; 9.43689570931383e-16 -1.339206523454095e-15 1.0000000000000007 -8.550505126287743e-16; -6.245004513516506e-16 -2.0122792321330962e-16 1.183061278035052e-16 1.0000000000000002],
    B = [0.09648289218436859 0.023497875751503007 0.0 0.0; 0.023497875751503007 0.045787575150300804 0.0 0.0; 0.0 0.0 0.0 0.0; 0.0 0.0 0.0 0.0]
    @test sqrt(A*B*A') ≈ A*B*A'
end
```

File: stdlib/LinearAlgebra/src/triangular.jl (hunk 4 above)
Comment (ssikdar1) [reply stdlib/LinearAlgebra/src/triangular.jl]: changed to this in faa95fd

File: stdlib/LinearAlgebra/test/triangular.jl (hunk 5 above)
Comment (ssikdar1) [reply stdlib/LinearAlgebra/test/triangular.jl]: addressed in faa95fd

**Comment by ssikdar1** [comment]
@stevengj Is anything I need to for the backport labels?

**Review by stevengj — APPROVED** [review]

**Comment by stevengj** [comment]
@ssikdar1, no, the backport labels are used in the release process for 1.3.x etcetera to know which commits to backport.

**Pull request merged**